cpp/src/io/json/nested_json_gpu.cu

```diff
@@ -0,0 +1,638 @@
+  {
+    return (stack_symbol == '{' || stack_symbol == '[')   ? fst::stack_op_type::PUSH
+           : (stack_symbol == '}' || stack_symbol == ']') ? fst::stack_op_type::POP
+                                                          : fst::stack_op_type::READ;
+  }
+};
+
+namespace detail {
+
+void get_stack_context(device_span<SymbolT const> d_json_in,
+                       SymbolT* d_top_of_stack,
+                       rmm::cuda_stream_view stream)
+{
+  constexpr std::size_t single_item = 1;
+
+  // Symbol that will represent empty-stack (i.e., that we're at the DOM root)
+  constexpr StackSymbolT root_symbol = '_';
+  // This can be any stack symbol from the stack alphabet that does not push onto stack
+  constexpr StackSymbolT read_symbol = 'x';
+
+  // Number of stack operations in the input (i.e., number of '{', '}', '[', ']' outside of quotes)
+  hostdevice_vector<SymbolOffsetT> d_num_stack_ops(single_item, stream);
+
+  // Sequence of stack symbols and their position in the original input (sparse representation)
+  rmm::device_uvector<StackSymbolT> d_stack_ops{d_json_in.size(), stream};
+  rmm::device_uvector<SymbolOffsetT> d_stack_op_indices{d_json_in.size(), stream};
+
+  // Prepare finite-state transducer that only selects '{', '}', '[', ']' outside of quotes
+  using ToStackOpFstT =
+    cudf::io::fst::detail::Dfa<StackSymbolT,
+                               static_cast<int32_t>(
+                                 to_stack_op::DFASymbolGroupID::NUM_SYMBOL_GROUPS),
+                               to_stack_op::DFA_STATES::TT_NUM_STATES>;
+  ToStackOpFstT json_to_stack_ops_fst{to_stack_op::symbol_groups,
+                                      to_stack_op::transition_table,
+                                      to_stack_op::translation_table,
+                                      stream};
+
+  // "Search" for relevant occurrence of brackets and braces that indicate the beginning/end
+  // structs/lists
+  json_to_stack_ops_fst.Transduce(d_json_in.begin(),
+                                  static_cast<SymbolOffsetT>(d_json_in.size()),
+                                  d_stack_ops.data(),
+                                  d_stack_op_indices.data(),
+                                  d_num_stack_ops.device_ptr(),
+                                  to_stack_op::start_state,
+                                  stream);
+
+  // Request temporary storage requirements
+  fst::sparse_stack_op_to_top_of_stack<StackLevelT>(
+    d_stack_ops.data(),
+    device_span<SymbolOffsetT>{d_stack_op_indices.data(), d_stack_op_indices.size()},
+    JSONToStackOp{},
+    d_top_of_stack,
+    root_symbol,
+    read_symbol,
+    d_json_in.size(),
+    stream);
+}
+
+void get_token_stream(device_span<SymbolT const> d_json_in,
+                      PdaTokenT* d_tokens,
+                      SymbolOffsetT* d_tokens_indices,
+                      SymbolOffsetT* d_num_written_tokens,
+                      rmm::cuda_stream_view stream)
```

Comment:
Typically, cudf APIs return the value that they are generating. I assume that the reason for passing `d_tokens*` arguments using pre-allocated pointers is to reuse the memory in cuio. If they will be allocated everytime, it's good to add memory resource as argument and allocate the required memory inside this function and return the values (as pair or tuple). 
@vuule could you comment about this?

Reply:
Hard to tell since we only have a unit test for now, but returning a tuple of `device_uvector`s is probably better. I don't think we need to pass `mr`, since this function only returns intermediate results in a `read_json` call.

That said, I'm fine with a TODO here and we can easily improve this later on.